### ISSUE TITLE
Misleading error message for `re.match()` on an Optional[str]

### ISSUE THREAD

**Issue by lengau** [microsoft/pyright#5058]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
If you pass a value that pyright has determined to be an `Optional[str]` to the `value` of `re.match`, it presents a misleading error message.

**To Reproduce**
```
(typing) lengau@penguin:~$ cat test.py 
from typing import Optional
import re

def match_thing(string: Optional[str]):
    return re.match(r"abc", string)
(typing) lengau@penguin:~$ pyright --version
pyright 1.1.306
(typing) lengau@penguin:~$ pyright test.py
/home/lengau/test.py
  /home/lengau/test.py:5:21 - error: Argument of type "Literal['abc']" cannot be assigned to parameter "pattern" of type "bytes | Pattern[bytes]" in function "match"
    Type "Literal['abc']" cannot be assigned to type "bytes | Pattern[bytes]"
      "Literal['abc']" is incompatible with "bytes"
      "Literal['abc']" is incompatible with "Pattern[bytes]" (reportGeneralTypeIssues)
  /home/lengau/test.py:5:29 - error: Argument of type "str | None" cannot be assigned to parameter "string" of type "ReadableBuffer" in function "match"
    Type "str | None" cannot be assigned to type "ReadableBuffer"
      Type "str" cannot be assigned to type "ReadableBuffer"
        "str" is incompatible with "ReadOnlyBuffer"
        "str" is incompatible with "bytearray"
        "str" is incompatible with "memoryview"
        "str" is incompatible with "array[Any]"
        "str" is incompatible with "mmap"
        "str" is incompatible with "_CData"
    ... (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 

```

**Expected behavior**
I expect to see the second error message, but not the first.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from typing import Optional
import re

def match_thing(string: Optional[str]):
    return re.match(r"abc", string)
```

**Comment by erictraut** [comment]
The `re.match` function is overloaded. It contains two overload signatures. When pyright performs overload matching, it first filters any signatures that do not match in arity (i.e. the number of arguments doesn't match the number of parameters) and keyword names. This produces a list of candidate overloads. It then evaluates the types of the argument expressions and attempts to find an overload whose signature types match. In cases where no overloads are successfully matched, pyright used to report a diagnostic message that said something like "no overloads match". That left users with no idea which arguments were incorrect. Based on user feedback, I changed the logic to always evaluate the argument list with the last candidate overload. (This is somewhat arbitrary, but it's a reasonable guess since the last overload is often the most general.) This approach allows pyright to evaluate each argument expression and emit diagnostics for each, which is often preferable to a more general "no overloads found" statement.

In the particular case you encountered, the technique doesn't work so well. However, I don't see a way to address it without significantly degrading the error messages in more common cases. So I think the right thing to do is leave it as it is.
